modules/plugin/arcsde/datastore/src/main/java/org/geotools/arcsde/data/ArcSDEQuery.java

```diff
@@ -591,69 +591,77 @@ public int calculateResultCount() throws IOException, UnsupportedOperationExcept
-                    IOException {
+    final Command<Integer> countCmd = new Command<Integer>() {
+        final boolean calcMasks = true; // use the spatial query to calculate
 
-                final SeQueryInfo queryInfo = qInfo;
+        // statistics.
+        final short searchOrder = SeQuery.SE_OPTIMIZE;
 
-                SeQuery query = new SeQuery(connection);
+        @Override
+        public Integer execute(ISession session, SeConnection connection)
+                throws SeException, IOException {
+            SeQuery query = new SeQuery(connection);
 
-                try {
-                    versioningHandler.setUpStream(session, query);
-
-                    if (spatialFilters != null && spatialFilters.length > 0) {
-
-                        final boolean calcMasks = true;// use the spatial query to calculate
-                        // statistics.
-                        final short searchOrder = SeQuery.SE_OPTIMIZE;
-                        query.setSpatialConstraints(searchOrder, calcMasks, spatialFilters);
-
-                        final SeDBMSInfo dbmsInfo = connection.getDBMSInfo();
-                        final boolean unsupported = versioningHandler != ArcSdeVersionHandler.NONVERSIONED_HANDLER
-                                && dbmsInfo.dbmsId == SeDBMSInfo.SE_DBMS_IS_ORACLE;
-
-                        if (unsupported) {
-                            LOGGER.fine("ArcSDE on Oracle can't calculate count statistics "
-                                    + "on versioned layers with spatial filters");
-                            /*
-                             * Despite the FeatureSource.getCount() contract saying it's ok to
-                             * return -1 if count is too expensive to calculate, the GeoServer
-                             * codebase is plagued of FeatureCollection.size() calls depending on
-                             * actual result counts or some operations don't work at all. return -1;
-                             */
-
-                            query.prepareQueryInfo(queryInfo);
-                            query.execute();
-                            int count = 0;
-                            while (query.fetch() != null) {
-                                count++;
-                            }
-                            return count;
-                        }
+            try {
+                versioningHandler.setUpStream(session, query);
+
+                if (hasSpatialFilter) {
+                    /*
+                     * Despite the FeatureSource.getCount() contract saying it's
+                     * ok to return -1 if count is too expensive to calculate,
+                     * the GeoServer codebase is plagued of
+                     * FeatureCollection.size() calls depending on actual result
+                     * counts or some operations don't work at all. return -1;
+                     */
```

Comment:
That comment was in the previous code.  I didn't try and see if the count
could be returned as -1 (which would improve performance even more - but
sounded like it might break things in ways which would not show up during
brief testing).  So I just left the comment in there.

It's easy to remove the comment.  I might even be able to try just
returning -1.  But I'm worried that might work for me, but leave a time
bomb ticking...

On Sat, Apr 25, 2015 at 1:48 AM, Andrea Aime <notifications@github.com>
wrote:

> In
> modules/plugin/arcsde/datastore/src/main/java/org/geotools/arcsde/data/ArcSDEQuery.java
> <https://github.com/geotools/geotools/pull/828#discussion_r29099080>:
>
> > -                            while (query.fetch() != null) {
> > -                                count++;
> > -                            }
> > -                            return count;
> > -                        }
> > +            try {
> > +                versioningHandler.setUpStream(session, query);
> > +
> > +                if (hasSpatialFilter) {
> > +                    /*
> > +                     * Despite the FeatureSource.getCount() contract saying it's
> > +                     * ok to return -1 if count is too expensive to calculate,
> > +                     * the GeoServer codebase is plagued of
> > +                     * FeatureCollection.size() calls depending on actual result
> > +                     * counts or some operations don't work at all. return -1;
> > +                     */
>
> Mind, there is no contract breackage here. FeatureSource.getCount() can
> return -1, featureCollection.size() cannot, it must scroll over the
> features in case a fast count cannot be performed.
> So, if you experienced problems, there must be some feature collection
> that is not abiding to its contract.
> I'd thus remove this comment.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geotools/geotools/pull/828/files#r29099080>.
>